lib/protein/adapters/amqp_adapter/server.ex

```diff
@@ -14,43 +14,66 @@ defmodule Protein.AMQPAdapter.Server do
   def init(opts) do
     Process.flag(:trap_exit, true)
     chan = connect(opts)
-    {:ok, {chan, opts, 0}}
+    state = %{channel: chan, opts: opts, consumers: []}
+    {:ok, state}
   end
 
-  def terminate(_reason, {_chan, _opts, running_processes_count}) do
-    wait_for_all_processes(running_processes_count)
+  def terminate(_reason, %{consumers: consumers}) do
+    wait_for_all_consumers(consumers)
   end
 
-  def wait_for_all_processes(0), do: :ok
+  def wait_for_all_consumers([]), do: :ok
 
-  def wait_for_all_processes(running_processes_count) do
+  def wait_for_all_consumers(consumers) do
     receive do
-      {:DOWN, _, :process, _, :normal} -> wait_for_all_processes(running_processes_count - 1)
-      _ -> wait_for_all_processes(running_processes_count)
+      {:DOWN, _, :process, down_pid, :normal} ->
+        present_consumers = Enum.filter(consumers, fn %{pid: pid} -> pid != down_pid end)
+        wait_for_all_consumers(present_consumers)
+
+      _ ->
```

Comment:
This is a little bit beyond my understanging of elixir: Is it safe to ignore all other messages?